selfie.c

```diff
@@ -7677,6 +7759,58 @@ void store_symbolic_memory(uint64_t vaddr, uint64_t val, char* sym, char* var, u
   symbolic_memory = sword;
 }
 
+void copy_symbolic_memory(uint64_t* from_context, uint64_t* to_context) {
+  uint64_t* sword;
+  uint64_t* sword_copy;
+  uint64_t* symbolic_memory_copy;
+  uint64_t* previous;
+
+  sword = get_symbolic_memory(from_context);
+
+  sword_copy           = (uint64_t*) 0;
+  symbolic_memory_copy = (uint64_t*) 0;
+  previous             = (uint64_t*) 0;
+
+  while (sword) {
+    sword_copy = allocate_symbolic_memory_word();
+
+    set_word_address(sword_copy, get_word_address(sword));
+    set_word_value(sword_copy, get_word_value(sword));
+    set_word_symbolic(sword_copy, get_word_symbolic(sword));
+    set_number_of_bits(sword_copy, get_number_of_bits(sword));
+
+    if (previous != (uint64_t*) 0)
+      set_next_word(previous, sword_copy);
+
+    if (symbolic_memory_copy == (uint64_t*) 0)
+      symbolic_memory_copy = sword_copy;
+
+    previous = sword_copy;
+    sword = get_next_word(sword);
+  }
+
+  set_next_word(sword_copy, 0);
+  set_symbolic_memory(to_context, symbolic_memory_copy);
+}
+
+void delete_word_from_symbolic_memory(uint64_t vaddr) {
+  uint64_t* sword;
+
+  sword = symbolic_memory;
+
+  while (sword) {
+    if (get_word_address(sword) == vaddr)
+      /* we indicate that this word has been deleted in order to avoid finding
```

Comment:
I remember that I solved this problem by simply adding new values of a word to the front of a list of values. Later, when looking for the value of a word I searched from the beginning of the list and then stopped when I found the first value. Would that work here as well?

Reply:
Also here. Is this true?

Reply:
I do not think so, but I am not quite sure yet. I am currently working on it and will tell you as soon as I am sure and have a good reasoning.

Reply:
In `merge_symbolic_store` I have to go through the symbolic memory of both contexts to merge words with the same virtual address. That means I would have to remember (at least in the outer loop) which words (or which virtual addresses) I have already looked at, so this would lead to something similar and I thought that it would be a bit clearer to mark such words as DELETED right from the beginning.

Reply:
I believe in that case it is easier and temporally and spatially more efficient to maintain the invariant that there is at most one entry per vaddr in any symbolic store. This means whenever you create an entry for a given vaddr you need to search the symbolic store. If you find an entry for vaddr just update that entry and you are done, thereby avoiding any duplicate entries. If you do not find an entry create a new entry in the store. The store is a singly-linked list with new entries added at the head of the list and existing entries searched by also starting at the head of the list. Despite quadratic complexity, this should facilitate fast symbolic storing with code that exhibits temporal locality (a store to a given vaddr is likely followed by another store to the same address) which most code does. Overall, the merge should then become very simple, right?

Reply:
Actually, I am not sure if this works but I am also not sure if the mark-as-deleted approach works. Symbolic stores are singly-linked list but they are partially shared among different contexts, at least in my original design. So, if we overwrite or mark entries as deleted in a given symbolic store we may change symbolic stores of other contexts as well, namely contexts that were created earlier. What we need is a merge of symbolic stores that avoids that.

Reply:
I suggest to neither overwrite nor mark entries as deleted in symbolic stores but instead implement a proper merge.

Reply:
In my version each context has its own symbolic store, so it is no longer shared. Since it may very well be that two contexts overwrite the 'same' variable differently, I think it is even necessary that each context has its own symbolic store. So the approach of overwriting existing variables works for sure and is also more elegant than marking as deleted. However, are you sure that a shared symbolic store is possible or a better solution, respectively?

Reply:
It may be important to keep algorithmic complexity down. I think, shared stores should work, yes, and may help with scalability. However, if your implementation is correct, then keep it for now. You can come back to this issue later. Then, I’d not copy symbolic stores at all but remember in each context the head pointer to the store whenever the context is branched into two contexts. That pointer marks the shared portion of the store. New stores can then be added to the store by updating existing entries but only in the unshared portion, that is, in the unshared portion there is at most one entry per vaddr. Merging (the unshared portion of) two stores is then easy.

Reply:
Ok, I will take care of the other things first and deal with this when the other things are done.

Reply:
I am a little unsure about this approach. Let us look at the following example: `c1` (context1) encounters a `beq` and creates `c2` , then both set a pointer to the end of the shared store and then insert their new stores into the unshared store. Now both `c1` and `c2` encounter another `beq` and create `c3` (from `c1`) and `c4` (from `c2`). The shared store between `c1` and `c3` is now different than the shared store between `c1` and `c2`, the same is true for the shared store between `c2` and `c4`. That means `c1` and `c2` have now different pointers than before. After (in the best case) `c1` is merged with `c3` (resulting in `c1`) and `c2` with `c4` (resulting in `c2`),  `c1` probably will be merged with `c2`. In that case, you would have to update the pointer again, so that it marks the end of the shared store correctly. So, you have to have at least one stack of these pointers. However, if a context makes an exit before it is merged (e.g. `c3` timeouts before it can be merged with `c1`), the pointer of the corresponding context (in this case `c1`) has to be adapted, in order to know which part of the store is shared and unshared in relation to the context it is going to be merged with. So, you would also need some additional information about how (more specifically by whom) the context was created. Or do I miss something and there is actually an easier way to do that?

Reply:
I am not sure if my idea of how to do this is correct. I can only say that each symbolic store is a singly-linked list of vaddr-to-value mappings. All symbolic stores together form a tree of such mappings. The tree structure corresponds to the control flow of the code with the beq instructions corresponding to the nodes in the tree. Whenever we merge any two symbolic contexts they form a `Y` with the lower part of the `Y` being the shared portion and the two upper parts of the `Y` being their respective unshared portions. The only thing to remember is where they split in the `Y`. Then, we need to compute a `|` by keeping the shared portion and merging the two unshared portions.

Reply:
But I can see that there may be problems merging contexts in arbitrary order, not sure.

Reply:
Yes, that is the problem, it cannot be guaranteed that all contexts are merged together in the 'right' order (since a context can timeout or exit before reaching its merge location).
But here is another idea: What if the whole symbolic store is shared among all contexts with each entry having an additional value that indicates to which context it belongs to (so the symbolic memory word would have an additional value)? New stores are just added at the top of the symbolic memory, without any overwriting or marking as deleted. So, branching into two contexts (if we copy the pointer mentioned later) and adding new stores would cost us constant time. The only downside is, that the store will grow relatively large. However, each context could have another pointer which always points to his last inserted value. That means, if we search for a value, we do not start looking from the top of the symbolic store, but just from the position the pointer points to. So, we could still take advantage of temporal locality. With this additional pointer, searching for values should have a (in most cases) an acceptable runtime. What do you think about that?

Reply:
Unfortunately, the idea I mentioned before (with the whole memory shared) doesn't work so easily. But I may have found a way to make your idea work. I'll have to formalize it a bit, I'll let you know.

Reply:
I have implemented the (partially) shared symbolic memory according to your idea with some modifications (marking the begin of the shared symbolic memory in the symbolic memory itself instead of using pointers in the contexts). The merging is actually more complicated now, because everything needs to be stored in the right 'place' and that changes with every merge/context. However, merging is also more efficient than before, as are the other operations (inserting, branching, searching). So, the (partially) shared memory does improve the performance by quite a lot!

I think, however, I have to work on the comments so that it becomes clearer what happens in the memory.